### PR TITLE
feat(Core/Maps): AreaBoundary 

### PR DESCRIPTION
**Pull request by IntelligentQuantum**
## Changes Proposed:
-  
-  

## Issues Addressed:

- Closes 

## SOURCE:

https://github.com/TrinityCore/TrinityCore/commit/2da458c56d024aac04468ce2af454a83ad790bf6
## Tests Performed:

- 
- 

## How to Test the Changes:

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Lynethris** [comment]
I passed many hours to travel different regions, and raid. I didn't find any issue with the world and the evade NPC. No error in worldserver observe. But I don't really know what to look for precisely. Maybe someone else can test too to be sure ?

**Comment by Nyeriah** [comment]
This PR implements area boundary reset modes for bosses. Right now bosses evade once they move past specific game objects (doors) without acknowledging coordinates. This PR implements boundaries so bosses reset once such boundaries are reached instead of using the door game objects,

**Comment by Lynethris** [comment]
Tested this night : 

- [X] Evade in BWL
- [X] Evade in BRD
- [X] Evade in Ulduar
- [X] Evade in UBRS

Resume: No big issue find. Just Vaelestraz (in BWL) can pass through the wall if the players go out the room.

**Comment by IntelligentQuantum** [comment]
> Tested this night :
> 
>     * [ ]  Evade in BWL : Just Vaelestraz can pass through the wall if the players go out the room (but seems not concern by this PR, right ?)

Hey tnx for test, can you test this part on clean ac please?

**Comment by Nyeriah** [comment]
I'm not certain what wall was tested but it's blizzlike to be able to move Vael out of his room

image

**Comment by IntelligentQuantum** [comment]
@Nyeriah So ready to merge or ... ?

**Comment by Nyeriah** [comment]
> @Nyeriah So ready to merge or ... ?

Up to @Nefertumm

File: src/server/game/AI/ScriptedAI/ScriptedCreature.cpp [hunk 1]
```diff
@@ -504,7 +504,7 @@ bool ScriptedAI::EnterEvadeIfOutOfCombatArea()
 
     EnterEvadeMode();
     return true;
-}
+}*/
```
Comment (Nefertumm) [inline src/server/game/AI/ScriptedAI/ScriptedCreature.cpp:507]: Why commenting? If it's not used anymore, just delete it

**Comment by Lynethris** [comment]
I have redone some quick tests  with :

- [X] Hummel Apothicary
- [X] Naxxramas
- [X] Zul Gurub
- [X] Hall of reflection

For me all seems ok. But I can do more tests if needed.

**Comment by Nefertumm** [comment]
Still build errors

**Comment by IntelligentQuantum** [comment]
@Nefertumm Done and i cant edit eluna repo so please fix eluna

File: src/server/game/AI/CreatureAI.cpp [hunk 2]
```diff
@@ -359,6 +361,44 @@ void CreatureAI::MoveBackwardsChecks() {
     me->GetMotionMaster()->MoveBackwards(victim, moveDist);
 }
 
+bool CreatureAI::IsInBoundary(Position const* who) const
+{
+    if (!_boundary)
+        return true;
+
+    if (!who)
+        who = me;
```
Comment (Winfidonarleyan) [inline src/server/game/AI/CreatureAI.cpp:370]: brackets

File: src/server/game/AI/CreatureAI.cpp [hunk 3]
```diff
@@ -359,6 +361,44 @@ void CreatureAI::MoveBackwardsChecks() {
...
+
+    return (CreatureAI::IsInBounds(*_boundary, who) != _negateBoundary);
+}
+
+bool CreatureAI::IsInBounds(CreatureBoundary const& boundary, Position const* pos)
+{
+    for (AreaBoundary const* areaBoundary : boundary)
+        if (!areaBoundary->IsWithinBoundary(pos))
+            return false;
```
Comment (Winfidonarleyan) [inline src/server/game/AI/CreatureAI.cpp:379]: brackets

File: src/server/game/Entities/Creature/Creature.cpp [hunk 4]
```diff
@@ -689,6 +689,17 @@ void Creature::Update(uint32 diff)
...
 
+                // periodic check to see if the creature has passed an evade boundary
+                if (IsAIEnabled && !IsInEvadeMode() && IsInCombat())
+                {
+                    if (diff >= m_boundaryCheckTime)
+                    {
+                        AI()->CheckInRoom();
+                        m_boundaryCheckTime = 2500;
+                    } else
+                        m_boundaryCheckTime -= diff;
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Creature/Creature.cpp:700]: codestyle

File: src/server/game/Instances/InstanceScript.cpp [hunk 5]
```diff
@@ -31,6 +31,12 @@
 #include "Spell.h"
 #include "WorldSession.h"
 
+BossBoundaryData::~BossBoundaryData()
+{
+    for (const_iterator it = begin(); it != end(); ++it)
+        delete it->boundary;
```
Comment (Winfidonarleyan) [inline src/server/game/Instances/InstanceScript.cpp:37]: brackets + just use range for

File: src/server/game/Maps/AreaBoundary.cpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        pos->GetPositionX() > _maxX ||
+        pos->GetPositionY() < _minY ||
+        pos->GetPositionY() > _maxY
+        );
+}
+
+// ---== CIRCLE ==---
+CircleBoundary::CircleBoundary(Position const& center, double radius, bool isInverted) :
+    AreaBoundary(isInverted), _center(center), _radiusSq(radius*radius) { }
+CircleBoundary::CircleBoundary(Position const& center, Position const& pointOnCircle, bool isInverted) :
```
Comment (Winfidonarleyan) [inline src/server/game/Maps/AreaBoundary.cpp:38]:
```suggestion

CircleBoundary::CircleBoundary(Position const& center, Position const& pointOnCircle, bool isInverted) :
```

File: src/server/game/Maps/AreaBoundary.cpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    AreaBoundary(isInverted), _center(center), _radiusSq(_center.GetDoubleExactDist2dSq(pointOnCircle)) { }
+bool CircleBoundary::IsWithinBoundaryArea(Position const* pos) const
```
Comment (Winfidonarleyan) [inline src/server/game/Maps/AreaBoundary.cpp:40]:
```suggestion

bool CircleBoundary::IsWithinBoundaryArea(Position const* pos) const
```

File: src/server/game/Maps/AreaBoundary.cpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+{
+    double offX = _center.GetDoublePositionX() - pos->GetPositionX();
+    double offY = _center.GetDoublePositionY() - pos->GetPositionY();
+    return offX*offX+offY*offY <= _radiusSq;
```
Comment (Winfidonarleyan) [inline src/server/game/Maps/AreaBoundary.cpp:44]:
```suggestion
    return offX * offX + offY * offY <= _radiusSq;
```

**Review by Winfidonarleyan — APPROVED** [review]

**Comment by Nyeriah** [comment]
Merge conflicts

**Comment by Nyeriah** [comment]
image

Eluna needs to be fixed 🤷

**Comment by IntelligentQuantum** [comment]
@Nyeriah Hmm done

**Comment by 55Honey** [comment]
Eluna buildfix is ready, thanks.

**Pull request merged**